src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor.go

```diff
@@ -1,43 +1,291 @@
+	workerAddr addr.Address,
+	peerId libp2p.PeerID,
+) addr.Address {
+	var msgSender addr.Address // TODO replace this
 
-// Actor
-func (spa *StoragePowerActor_I) ReportConsensusFault(slasherAddr addr.Address, faultType ConsensusFaultType, proof []block.Block) {
-	panic("TODO")
+	// TODO: anything to check here?
+	newMiner := &PowerTableEntry_I{
+		ActivePower_:            block.StoragePower(0),
+		InactivePower_:          block.StoragePower(0),
+		AvailableBalance_:       actor.TokenAmount(0),
+		LockedPledgeCollateral_: actor.TokenAmount(0),
+	}
+	spa.PowerTable()[msgSender] = newMiner
 
-	// Use EC's IsValidConsensusFault method to validate the proof
-	// slash block miner's pledge collateral
-	// reward slasher
+	// TODO: commit state
+
+	// TODO: call constructor of StorageMinerActor
+	// store ownerAddr and workerAddr there?
+	// and return StorageMinerActor address
+
+	var smAddress addr.Address
+	return smAddress
 }
 
-func (spa *StoragePowerActor_I) ReportUncommittedPowerFault(cheaterAddr addr.Address, numSectors util.UVarint) {
-	panic("TODO")
-	// Quite a bit more straightforward since only called by the cron actor (ie publicly verified)
+func (spa *StoragePowerActor_I) RemoveStorageMiner(address addr.Address) {
+	isMinerVerified := spa.verifyMiner(address)
+	if !isMinerVerified {
+		// TODO: proper throw
+		panic("TODO")
+	}
 
-	// slash cheater pledge collateral accordingly based on num sectors faulted
+	if (spa.PowerTable()[address].ActivePower() + spa.PowerTable()[address].InactivePower()) > 0 {
+		// TODO: proper throw
+		panic("TODO")
+	}
+
+	delete(spa.PowerTable(), address)
+
+	// TODO: commit state
 }
 
-func (spa *StoragePowerActor_I) CommitPledgeCollateral(deals []deal.StorageDeal) {
+func (spa *StoragePowerActor_I) verifyMiner(address addr.Address) bool {
+	// TODO: anything else to check?
+	// TODO: check miner pledge collateral balances?
+	// TODO: decide on what should be checked here
+	_, found := spa.PowerTable()[address]
+	if !found {
+		return false
+	}
+	return true
+}
 
-	panic("TODO")
-	// check that based on deals (ie sector sizes and num sectors) miner has enough associated balance in the storage miner wallet
-	// pledge and associate
+func (spa *StoragePowerActor_I) GetTotalPower() block.StoragePower {
+	totalPower := block.StoragePower(0)
+	for _, miner := range spa.PowerTable() {
+		totalPower = totalPower + miner.ActivePower() + miner.InactivePower()
+	}
+	return totalPower
+}
+
+func (spa *StoragePowerActor_I) GetPledgeCollateralReq(power block.StoragePower) actor.TokenAmount {
+	// TODO: Implement
+	return actor.TokenAmount(0)
+}
+
+func (spa *StoragePowerActor_I) IsPledgeCollateralSatisfied() bool {
+	var msgSender addr.Address // TODO replace this
+
+	powerEntry := spa.PowerTable()[msgSender]
+	pledgeCollateralRequired := spa.GetPledgeCollateralReq(powerEntry.ActivePower() + powerEntry.InactivePower())
+
+	if pledgeCollateralRequired < powerEntry.LockedPledgeCollateral() {
+		return true
```

Comment:
leaning towards no, this function just checks to make sure that miner has the collateral required. To enable unlocking, perhaps we can have `UnlockPledgeCollateral` instead of `unlockPledgeCollateral`?

Reply:
Alternatively, this function does not change state and miner needs to call lock and unlock to change state for it

Reply:
Fine for now, but it is weird to leave more than enough locked. lets not block PR and revisit later